frame/proxy/src/lib.rs

```diff
@@ -0,0 +1,273 @@
+			force_proxy_type: Option<T::ProxyType>,
+			call: Box<<T as Trait>::Call>
+		) {
+			let who = ensure_signed(origin)?;
+			ensure!(T::IsCallable::filter(&call), Error::<T>::Uncallable);
+			let (_, proxy_type) = Proxies::<T>::get(&real).0.into_iter()
+				.find(|x| &x.0 == &who && force_proxy_type.as_ref().map_or(true, |y| &x.1 == y))
+				.ok_or(Error::<T>::NotProxy)?;
+			match call.is_sub_type() {
+				Some(Call::add_proxy(_, ref pt)) | Some(Call::remove_proxy(_, ref pt)) =>
+					ensure!(&proxy_type == pt, Error::<T>::NoPermission),
+				_ => (),
+			}
+			ensure!(proxy_type.filter(&call), Error::<T>::Unproxyable);
+			let e = call.dispatch(frame_system::RawOrigin::Signed(real).into());
+			Self::deposit_event(Event::ProxyExecuted(e.map(|_| ()).map_err(|e| e.error)));
+		}
+
+		/// Register a proxy account for the sender that is able to make calls on its behalf.
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// Parameters:
+		/// - `proxy`: The account that the `caller` would like to make a proxy.
+		/// - `proxy_type`: The permissions allowed for this proxy account.
+		///
+		/// # <weight>
+		/// P is the number of proxies the user has
+		/// - Base weight: 17.48 + .176 * P µs
+		/// - DB weight: 1 storage read and write.
+		/// # </weight>
+		#[weight = T::DbWeight::get().reads_writes(1, 1)
+				.saturating_add(18 * WEIGHT_PER_MICROS)
+				.saturating_add((200 * WEIGHT_PER_NANOS).saturating_mul(T::MaxProxies::get().into()))
+		]
+		fn add_proxy(origin, proxy: T::AccountId, proxy_type: T::ProxyType) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			Proxies::<T>::try_mutate(&who, |(ref mut proxies, ref mut deposit)| {
+				ensure!(proxies.len() < T::MaxProxies::get() as usize, Error::<T>::TooMany);
+				let typed_proxy = (proxy, proxy_type);
+				let i = proxies.binary_search(&typed_proxy).err().ok_or(Error::<T>::Duplicate)?;
+				proxies.insert(i, typed_proxy);
+				let new_deposit = T::ProxyDepositBase::get()
+					+ T::ProxyDepositFactor::get() * (proxies.len() as u32).into();
+				if new_deposit > *deposit {
+					T::Currency::reserve(&who, new_deposit - *deposit)?;
+				} else if new_deposit < *deposit {
+					T::Currency::unreserve(&who, *deposit - new_deposit);
+				}
+				*deposit = new_deposit;
+				Ok(())
+			})
+		}
+
+		/// Unregister a proxy account for the sender.
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// Parameters:
+		/// - `proxy`: The account that the `caller` would like to remove as a proxy.
+		/// - `proxy_type`: The permissions currently enabled for the removed proxy account.
+		///
+		/// # <weight>
+		/// P is the number of proxies the user has
+		/// - Base weight: 14.37 + .164 * P µs
+		/// - DB weight: 1 storage read and write.
+		/// # </weight>
+		#[weight = T::DbWeight::get().reads_writes(1, 1)
+				.saturating_add(14 * WEIGHT_PER_MICROS)
+				.saturating_add((160 * WEIGHT_PER_NANOS).saturating_mul(T::MaxProxies::get().into()))
```

Comment:
Good candidate for  refund as we're charging as if `P == T::MaxProxies::get()`

Reply:
yeah, though with the max only being 32, i suspect it's insubstantial.